Refactors how bots scan for stuff

JohnFulpWillard opened the pull request:
## About The Pull Request

Refactors how bots scan for stuff and also improves Cleanbots' code.

This started off as a Cleanbot refactor but I got too-deep into how it scans for stuff, so I decided to instead focus my PR on that first since it's something I want to tread carefully with.

removes oldloc var from Cleanbot, Medibot and Floorbot, because it was all unused.

I didn't even notice how bad this was before, because Secbots use their own snowflake thing instead of scanning like all the other bots do, which sounds like a good future PR.

Video of it in-game:
https://cdn.discordapp.com/attachments/932297735967752252/971198188914634802/2022-05-03_19-54-31.mp4

## Why It's Good For The Game

Bots seems to react more efficiently than they did previously from my testing, and it optimizes the code a bit.
Continuation of my bot code refactor PRs, I was hoping to finally get around to cleanbot, and this scan() proc was what scared me from doing it for the longest time.

## Changelog

Not needed

JohnFulpWillard committed:
refactors bot scanning

Refactors how bots scan for stuff and also improves Cleanbots' code.
This started off as a Cleanbot refactor but I got too-deep into how it scans for stuff, so I decided to instead focus my PR on that first since it's something I want to tread carefully with.

removes oldloc from Cleanbot, Medibot and Floorbot, because it was all unused.

Palpetinus commented:
I hope this fixes the issue that Cleanbots lock up completely if they see something to clean on a tile they can't get to.

JohnFulpWillard committed:
removes useless comment

JohnFulpWillard committed:
update docs

MrMelbert commented on code/modules/mob/living/simple_animal/bot/bot.dm:
```diff
@@ -482,53 +482,54 @@
-	if(!istype(scan, scan_type)) //Check that the thing we found is the type we want!
-		return FALSE //If not, keep searching!
-	if((REF(scan) in ignore_list) || (scan == old_target)) //Filter for blacklisted elements, usually unreachable or previously processed oness
-		return FALSE
+			continue //The current element failed assessment, move on to the next.
```
"else continue" at the end of the loop?

if it's intentional to be explicit feel free to ignore

MrMelbert commented on code/modules/mob/living/simple_animal/bot/bot.dm:
```diff
@@ -482,53 +482,54 @@
+			if(final_result)
+				return final_result
+
+		for(var/turf/scanned_turfs as anything in shuffle(view(scan_range, src))-adjacent) //Search for something in range, minus what we already checked.
```
```suggestion
		for(var/turf/scanned_turfs as anything in shuffle(view(scan_range, src)) - adjacent) //Search for something in range, minus what we already checked.
```

MrMelbert commented on code/modules/mob/living/simple_animal/bot/cleanbot.dm:
```diff
@@ -217,39 +249,38 @@
+	if(!target)
+		var/list/scan_targets = list(
+			/obj/effect/decal/cleanable,
+			/obj/effect/decal/remains,
+		)
 
-	if(!target && pests) //Search for pests to exterminate first.
-		target = scan(/mob/living/simple_animal)
+		if(bot_cover_flags & BOT_COVER_EMAGGED) // When emagged, target humans who slipped on the water and melt their faces off
+			scan_targets += list(/mob/living/carbon)
+		if(pests)
+			scan_targets += list(/mob/living/simple_animal)
+		if(trash)
+			scan_targets += list(
+				/obj/item/trash,
+				/obj/item/food/deadmouse,
+			)
```
Doesn't this change behavior? 

In the past it would scan for carbons, then simple animals, then cleanables and trash and stuff.
Now, it scans for cleanables, then carbons.

JohnFulpWillard replied:
It is, I’ll add it to the changelog

JohnFulpWillard replied:
Actually I’ll change it again later, I’ll draft this for now

MrMelbert commented on code/modules/mob/living/simple_animal/bot/bot.dm:
```diff
@@ -482,53 +482,54 @@
-		if(isturf(scan_type)) //If we're lookeing for a turf we can just run the checks directly!
-			var/final_result = checkscan(scan,scan_type,old_target)
+
+	for(var/scan_type_list in scan_types)
```
is scan_types a list of lists of typepaths, or just a list of typepaths, this var name confuses me

JohnFulpWillard replied:
A list of typepaths, scan_type_list was meant to say we’re going through it now

JohnFulpWillard committed:
Update code/modules/mob/living/simple_animal/bot/bot.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

JohnFulpWillard committed:
Removes useless continue

JohnFulpWillard committed:
improves cleanbot emag functions

Emagged cleanbots scan for humans first above all else.
Also removes the weird for() loop to check all carbons in loc

tralezab reviewed (approved):
with that weird thing melb noticed out of the way, looks good mane
